org/mozilla/jss/crypto/KBKDFDerivedKey.java

```diff
@@ -0,0 +1,107 @@
+package org.mozilla.jss.crypto;
+
+import java.lang.IllegalArgumentException;
+import java.security.InvalidKeyException;
+import java.util.Arrays;
+import java.util.ArrayList;
+import javax.crypto.SecretKey;
+import javax.crypto.SecretKeyFactory;
+
+import org.mozilla.jss.pkcs11.attrs.CKAttribute;
+import org.mozilla.jss.util.NativeEnclosure;
+
+/**
+ * Class for supporting additional derived keys from PKCS#11 v3.0's
+ * SP800-108 KBKDF implementation.
+ */
+public class KBKDFDerivedKey extends NativeEnclosure {
+    private CKAttribute[] attrs;
+    private long handle;
+
+    public KBKDFDerivedKey() {}
+
+    public KBKDFDerivedKey(CKAttribute[] attrs) throws IllegalArgumentException {
+        setAttributes(attrs);
+    }
+
+    public void setAttributes(CKAttribute[] attrs) throws IllegalArgumentException {
+        if (attrs == null) {
+            String msg = "Expected populated array of CKAttributes, but ";
+            msg += "was null.";
+            throw new IllegalArgumentException(msg);
+        }
+        this.attrs = attrs;
+    }
+
+    public void addAttribute(CKAttribute attr) throws IllegalArgumentException {
+        if (attr == null) {
+            String msg = "Expected non-null CKAttribute, but was null.";
+            throw new IllegalArgumentException(msg);
+        }
+
+        ArrayList<CKAttribute> arl;
+        if (attrs == null) {
+            arl = new ArrayList<CKAttribute>(1);
+        } else {
+            arl = new ArrayList<CKAttribute>(Arrays.asList(attrs));
+        }
+        arl.add(attr);
+
+        this.attrs = arl.toArray(new CKAttribute[0]);
```

Comment:
Minor. Generally the array passed to `list.toArray()` should be created with `list.size()` elements:
```
this.attrs = arl.toArray(new CKAttribute[arl.size()]);
```